Redecomp sparse matrix transfers

ebchin opened the pull request:
Implements sparse matrix transfers using redecomp.  Currently supports L2 spaces of order 0 and vdim = 1 (1 value per element).

ebchin committed:
initial commit of quad space matrix xfer

ebchin committed:
add QuadratureMatrixTransfer

ebchin committed:
document BuildParentElementRankOffsets

ebchin committed:
fix includes

jamiebramwell commented:
I assume this is to help solve @kswartz92 's filter problem?

ebchin commented:
> I assume this is to help solve @kswartz92 's filter problem?

That's correct

ebchin committed:
add optional ghost length

ebchin committed:
update cmake

kswartz92 committed:
Merge branch 'redecomp_elem_matrix_xfer' of github.com:LLNL/Tribol into redecomp_elem_matrix_xfer

ebchin committed:
Merge branch 'develop' into redecomp_elem_matrix_xfer

kswartz92 committed:
Merge branch 'redecomp_elem_matrix_xfer' of github.com:LLNL/Tribol into redecomp_elem_matrix_xfer

kswartz92 committed:
adding calculation of filter kernel

ebchin committed:
bugfixes

ebchin committed:
change quad space to fe space; do some renaming

kswartz92 committed:
testing filter

kswartz92 committed:
Merge branch 'redecomp_elem_matrix_xfer' of github.com:LLNL/Tribol into redecomp_elem_matrix_xfer

kswartz92 committed:
testing filter in parallel

ebchin committed:
use the parent mesh in hypreparmatrix csr

ebchin committed:
fix order of comm and update column map key

ebchin committed:
sorting offd column map

ebchin committed:
unique redecomp visit names

ebchin committed:
add test skeleton

kswartz92 committed:
implemented 2D and 3D density filter tests

ebchin committed:
Merge branch 'develop' into redecomp_elem_matrix_xfer

ebchin committed:
add filter radius to redecomp mesh ctor

kswartz92 committed:
adding comment and removing debug code

kswartz92 committed:
adding documentation to test

ebchin committed:
more comments and cleanup

ebchin commented:
@jamiebramwell and @srwopschall this one is ready for review. Thanks!

ebchin committed:
example fix

ebchin committed:
Merge branch 'develop' into redecomp_elem_matrix_xfer

jamiebramwell commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,306 @@
+    "MPI Communicator must match in test and trial spaces.");
+
+  // one value per element for initial implementation
+  SLIC_ERROR_ROOT_IF(!parent_test_space_.IsDGSpace() || !parent_trial_space_.IsDGSpace() ||
```
@kswartz92 , is there a need to make this more general than L2 order 0 in the near term?

kswartz92 replied:
I don't think so, L2 order 0 will cover most use cases.

jamiebramwell commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,306 @@
+      while (test_elem_id == test_elem_offsets[test_parent_rank + 1])
+      {
+        ++test_parent_rank;
+        test_ghost_ct = 0;
```
Why do you have to keep setting this to zero?

jamiebramwell commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,306 @@
+
+    // test space = row space
+    auto test_parent_rank = 0;
+    auto test_ghost_ct = 0;
```
What is the `test_ghost_ct`?

jamiebramwell commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,306 @@
+      }
+      // trial space = col space
+      auto trial_parent_rank = 0;
+      for (int Ij{src_I[i]}; Ij < src_I[i+1]; ++Ij)
```
I think some words explaining this algorithm would be welcome.

srwopschall replied:
@ebchin - Per @jamiebramwell 's comment, this is a complex algorithm that relies on an inherent understanding or familiarity with CSR arrays, some underlying knowledge of the structure of a hypre par matrix, and the mpi communication. Per her comment, please review for any extra explanation that may help elaborate on this algorithm.

ebchin committed:
additional comments

ebchin commented:
Thanks for taking a look @jamiebramwell, I believe I've addressed your comments in the latest commit.  @srwopschall, do you mind taking a quick look before merging?

ebchin committed:
Merge branch 'develop' into redecomp_elem_matrix_xfer

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,321 @@
+  );
+}
+
+const RedecompMesh& SparseMatrixTransfer::getRedecompMesh(const mfem::FiniteElementSpace& quadrature_space)
```
rename input argument. Not a quadrature space.

srwopschall commented on src/redecomp/RedecompMesh.cpp:
```diff
@@ -125,13 +127,18 @@ RedecompMesh::RedecompMesh(
 {
+  if (ghost_length < 0.0)
+  {
+    ghost_length = 1.25 * MaxElementSize(parent_, partitioner.getMPIUtility());
```
Add comment as to what this is calculating or the rationale behind this choice.

srwopschall commented on src/tests/redecomp_sparsematrix.cpp:
```diff
@@ -0,0 +1,223 @@
+  MPI_Barrier(MPI_COMM_WORLD);
+}
+
+INSTANTIATE_TEST_SUITE_P(redecomp, SparseMatrixTest, testing::Values(
```
Let's double check test runtime in CI. Let's double check that each refinement level is in fact testing something different. Perhaps refinement 1 and 2 are not testing anything new, but 3 guarantees zero element redecomp meshes, etc.

srwopschall commented on src/examples/element_matrix_redecomp.cpp:
```diff
@@ -4,7 +4,7 @@
- * @file matrix_redecomp.cpp
+ * @file element_matrix_redecomp.cpp
  *
  * @brief Demonstrates moving a matrix using redecomp::MatrixTransfer
```
Consider "Demonstrates moving element matrices using redecomp::MatrixTransfer"; this may be a little more precise since now we have multiple definitions of what "a matrix" is.

srwopschall commented on src/examples/sparse_matrix_redecomp.cpp:
```diff
@@ -0,0 +1,242 @@
+/**
+ * @file sparse_matrix_redecomp.cpp
+ *
+ * @brief Demonstrates moving a matrix using redecomp::SparseMatrixTransfer
```
Consider "Demonstrates moving a sparse matrix using redecomp::SparseMatrixTransfer"; it's a little pedantic, but more precise since we have multiple matrix transfers now.

srwopschall commented on src/examples/sparse_matrix_redecomp.cpp:
```diff
@@ -0,0 +1,242 @@
+ * Demonstrates use of redecomp::SparseMatrixTransfer to move a symmetric mfem::SparseMatrix whose rows and columns are
+ * associated with a mfem::FiniteElementSpace on a redecomp::RedecompMesh to its representation as an
+ * mfem::HypreParMatrix on the associated mfem::ParMesh and mfem::ParFiniteElementSpace. The row and column finite
+ * element space are the same in this example, but redecomp::SparseMatrixTransfer supports rectangular matrices with
```
spaces

srwopschall commented on src/examples/sparse_matrix_redecomp.cpp:
```diff
@@ -0,0 +1,242 @@
+ *   - mpirun -np 4 {build_dir}/examples/sparse_matrix_redecomp_ex -d 3 -r 2 -R 0.1
+ *
+ * @details This example relies on using a cartesian mesh with uniform element size to obtain an analytical definition
+ * of the filtered field. Do not change the mesh to anything other than a square without adjusting the analytical
```
square/cube?

srwopschall commented on src/redecomp/RedecompMesh.cpp:
```diff
@@ -125,13 +127,18 @@ RedecompMesh::RedecompMesh(
+  double ghost_length
 ) const
 {
+  if (ghost_length < 0.0)
```
Out of curiosity, what would happen if ghost length was 0.0? Or if ghost length was some small positive number that wasn't the 'length' of adjacent ghost elements? Maybe we guard for this possibility?

ebchin replied:
This should be fine -- there would be no ghost elements if the ghost length is 0.0.  Created a few new constructors so the ghost length based on the max element size isn't based on the value of the ghost length passed in.

srwopschall commented on src/redecomp/RedecompMesh.hpp:
```diff
@@ -148,9 +150,14 @@ class RedecompMesh : public mfem::Mesh
    * 
    * @param partitioner Method of partitioning the elements
    * @param n_parts Number of parts to partition the mesh into
+   * @param ghost_length Length around the domain border to include ghost elements
```
Minor, but it is really more of a "fuzz" width beyond the current domain border. Length around the domain border makes it sound like you only include ghost elements on a portion of the domain border.

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,321 @@
+  redecomp_trial_mesh_ { getRedecompMesh(redecomp_trial_space_) }
+{
+  SLIC_ERROR_ROOT_IF(static_cast<const mfem::Mesh*>(&redecomp_test_mesh_.getParent()) != parent_test_space_.GetMesh(),
+    "The parent test quadrature space mesh must be linked to the test Redecomp mesh.");
```
Change quadrature space to finite element space?

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,321 @@
+  SLIC_ERROR_ROOT_IF(static_cast<const mfem::Mesh*>(&redecomp_test_mesh_.getParent()) != parent_test_space_.GetMesh(),
+    "The parent test quadrature space mesh must be linked to the test Redecomp mesh.");
+  SLIC_ERROR_ROOT_IF(static_cast<const mfem::Mesh*>(&redecomp_trial_mesh_.getParent()) != parent_trial_space_.GetMesh(),
+    "The parent trial quadrature space mesh must be linked to the trial Redecomp mesh.");
```
Quadrature to finite element

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,321 @@
+  {
+    cmap[cmap_val.second] = cmap_val.first;
+  }
+  // compute j_diag and j_offd, the columns for each row (row offsets given by i).  will not be sorted.
```
populate vs. compute? Maybe elaborate on "will not be sorted".

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,321 @@
+    i_diag[i] = 0;
+    i_offd[i] = 0;
+  }
+  // maps from global j to order of appearance in the loops below
```
Consider clarity of comment. "in the loops below" isn't super clear. Global j may be hard to parse for someone not immediately refreshed on CSR data. This is complex code to be sure, but I'm not sure that this comment enhances understanding of the next handful of lines.

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,321 @@
+      }
+    }
+  }
+  // sum up i_diag and i_offd (completing their definition)
```
Consider clarity of comment. completing their definition is a little unclear.

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.cpp:
```diff
@@ -0,0 +1,321 @@
+const RedecompMesh& SparseMatrixTransfer::getRedecompMesh(const mfem::FiniteElementSpace& quadrature_space)
+{
+  auto redecomp_mesh = dynamic_cast<const RedecompMesh*>(quadrature_space.GetMesh());
+  SLIC_ERROR_ROOT_IF(redecomp_mesh == nullptr, "The quadrature space must have a Redecomp mesh.");
```
Rename quadrature space.

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.hpp:
```diff
@@ -0,0 +1,120 @@
+  /**
+   * @brief Build vector of parent element offsets for each MPI rank
+   * 
+   * @param redecomp_mesh RedecompMesh whose parent linked mesh the element offset vector is desired
```
Wording is confusing.

srwopschall commented on src/redecomp/transfer/SparseMatrixTransfer.hpp:
```diff
@@ -0,0 +1,120 @@
+  const mfem::FiniteElementSpace& redecomp_trial_space_;
+
+  /**
+   * @brief Test RedecompMesh
```
Not strictly required, but consider a note somewhere indicating that since you support rectangular sparse matrices you need to have two redecomp meshes. Up to now the paradigm has only been one on-rank redecomp mesh for a given redecomposition. It wouldn't hurt elaborating on why, even if internally, two are required.

srwopschall commented on src/tests/redecomp_sparsematrix.cpp:
```diff
@@ -0,0 +1,223 @@
+namespace redecomp {
+
+/**
+ * @brief This test transfers an mfem::SparseMatrix whose rows and colums are associated with an order-0 L2 field on a
```
Mentione if this is symmetric like the example.

srwopschall commented on src/tests/redecomp_sparsematrix.cpp:
```diff
@@ -0,0 +1,223 @@
+      }
+    }
+
+    W_redecomp.Finalize();
```
Consider noting that this is required in order to interface with the redecomp capability.

srwopschall reviewed (approved):
@ebchin - This looks excellent. I've made minor clarity comments that require addressing.

ebchin committed:
Merge branch 'develop' into redecomp_elem_matrix_xfer

ebchin committed:
Merge branch 'develop' into redecomp_elem_matrix_xfer

ebchin committed:
updated documentation and naming fixes

ebchin commented:
Thanks for the review @srwopschall.  I've addressed your comments in the latest commits.
